src/umpire/strategy/ResourceAwarePool.cpp

```diff
@@ -0,0 +1,549 @@
+      std::size_t bytes_to_use{(m_actual_bytes == 0) ? m_first_minimum_pool_allocation_size
+                                                     : m_next_minimum_pool_allocation_size};
+
+      std::size_t size{(rounded_bytes > bytes_to_use) ? rounded_bytes : bytes_to_use};
+
+      UMPIRE_LOG(Debug, "Allocating new chunk of size " << size);
+
+      void* ret{nullptr};
+      try {
+#if defined(UMPIRE_ENABLE_BACKTRACE)
+        {
+          umpire::util::backtrace bt;
+          umpire::util::backtracer<>::get_backtrace(bt);
+          UMPIRE_LOG(Info, "actual_size:" << (m_actual_bytes + rounded_bytes) << " (prev: " << m_actual_bytes << ") "
+                                          << umpire::util::backtracer<>::print(bt));
+        }
+#endif
+        ret = aligned_allocate(size); // Will Poison
+      } catch (...) {
+        UMPIRE_LOG(Error,
+                   "Caught error allocating new chunk, giving up free chunks and "
+                   "retrying...");
+        release();
+        try {
+          ret = aligned_allocate(size); // Will Poison
+          UMPIRE_LOG(Debug, "memory reclaimed, chunk successfully allocated.");
+        } catch (...) {
+          UMPIRE_LOG(Error, "recovery failed.");
+          throw;
+        }
+      }
+
+      m_actual_bytes += size;
+      m_releasable_bytes += size;
+      m_releasable_blocks++;
+      m_total_blocks++;
+      m_actual_highwatermark = (m_actual_bytes > m_actual_highwatermark) ? m_actual_bytes : m_actual_highwatermark;
+
+      void* chunk_storage{m_chunk_pool.allocate()};
+      chunk = new (chunk_storage) Chunk{ret, size, size, r};
+    } else {
+      chunk = (*best).second;
+      chunk->m_resource = r;
+      m_free_map.erase(best);
+    }
+  }
+
+  UMPIRE_LOG(Debug, "Using chunk " << chunk << " with data " << chunk->data << " and size " << chunk->size
+                                   << " for allocation of size " << rounded_bytes);
+
+  if ((chunk->size == chunk->chunk_size) && chunk->free) {
+    m_releasable_bytes -= chunk->chunk_size;
+    m_releasable_blocks--;
+  }
+
+  void* ret = chunk->data;
+  m_used_map.insert(std::make_pair(ret, chunk));
+
+  chunk->free = false;
+
+  if (rounded_bytes != chunk->size) {
+    std::size_t remaining{chunk->size - rounded_bytes};
+    UMPIRE_LOG(Debug, "Splitting chunk " << chunk->size << "into " << rounded_bytes << " and " << remaining);
+
+    void* chunk_storage{m_chunk_pool.allocate()};
+    Chunk* split_chunk{new (chunk_storage)
+                           Chunk{static_cast<char*>(ret) + rounded_bytes, remaining, chunk->chunk_size, r}};
+    auto old_next = chunk->next;
+    chunk->next = split_chunk;
+    split_chunk->prev = chunk;
+    split_chunk->next = old_next;
+
+    if (split_chunk->next)
+      split_chunk->next->prev = split_chunk;
+
+    chunk->size = rounded_bytes;
+    split_chunk->size_map_it = m_free_map.insert(std::make_pair(remaining, split_chunk));
+    split_chunk->free = true;
```

Comment:
Is this guaranteed to be free? Or could it still be pending?

Reply:
I think the split_chunk is meant to break off or split off a piece of the chunk that is too big given what's absolutely needed for the allocation. So like the extra room. Since it's a new chunk being created, it will definitely be free. This logic is also part of the QuickPool

Reply:
Thanks for explaining.